pkg/controller/releasemanager/releasemanager_controller.go

```diff
@@ -0,0 +1,407 @@
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to secondary resource Incarnations and Clusters and requeue the owner ReleaseManager
+	err = c.Watch(&source.Kind{Type: &picchuv1alpha1.Incarnation{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &picchuv1alpha1.ReleaseManager{},
+	})
+	if err != nil {
+		return err
+	}
+	err = c.Watch(&source.Kind{Type: &picchuv1alpha1.Cluster{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &picchuv1alpha1.ReleaseManager{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileReleaseManager{}
+
+// ReconcileReleaseManager reconciles a ReleaseManager object
+type ReconcileReleaseManager struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a ReleaseManager object and makes changes based on the state read
+// and what is in the ReleaseManager.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileReleaseManager) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ReleaseManager")
+
+	// Fetch the ReleaseManager instance
+	instance := &picchuv1alpha1.ReleaseManager{}
+	if err := r.client.Get(context.TODO(), request.NamespacedName, instance); err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	cluster := &picchuv1alpha1.Cluster{}
+	key := client.ObjectKey{request.Namespace, instance.Spec.Cluster}
+	if err := r.client.Get(context.TODO(), key, cluster); err != nil {
+		return reconcile.Result{}, err
+	}
+	incarnationList := &picchuv1alpha1.IncarnationList{}
+	labelSelector := client.MatchingLabels(map[string]string{
+		"medium.build/app":     instance.Spec.App,
+		"medium.build/cluster": instance.Spec.Cluster,
+	})
+	if err := r.client.List(context.TODO(), labelSelector, incarnationList); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	remoteClient, err := utils.RemoteClient(r.client, cluster)
+	if err != nil {
+		reqLogger.Error(err, "Failed to create remote client", "Cluster.Key", key)
+		return reconcile.Result{}, err
+	}
+	reqLogger.Info("Created remote client", "Cluster.Key", key)
+	manager := ReleaseManager{
```

Comment:
> `picchuv1alpha1.ReleaseManager`
> `ReleaseManager`

wowza this is confusing. is there another name we can use for the resource?

Reply:
sorry, noticed this late. I'll think about it